### ISSUE TITLE
Support for Variables Viewer from Jupyter Extension

### ISSUE THREAD

**Issue by DonJayamanne** [julia-vscode/julia-vscode#2020]
The Jupyter extension today has a variable viewer for display of variables in kernel.
As you execute code in a kernel, this gets updated and we have the ability to open external editors/viewers for some of the data (current that's hardcoded).

Here's what it looks like and what we're hoping it would look like with the new VSCode API https://github.com/microsoft/vscode/issues/118115#issuecomment-794540781 (second screnshot is the proposed looknfeel of the new viewer, using the new VSC table api).

Is this something that you'd be interested in.
FYI - Today we have extensibility points for other extensions to provide variables for other languages, however its limited to Jupyter kernels. I.e. only kernels that Jupyter kernels that the Jupyter extension starts.

However if this looks interesting we can consider some extensibility points to expose this to the Julia VS Code kernel. 
Basically when you run a cell agianst the Julia kernel, the variable viewer would display the current variables in the kernel. As you execute code, the list will get updated.
Optionally we can have custom actions against each item, allowing users to open different viewers for each variable.

**Comment by pfitzseb** [comment]
Just for reference, this is our current tree-based variables viewer:
![image](https://user-images.githubusercontent.com/6735977/110915839-74628700-8318-11eb-8045-6253164cbb00.png)

IMHO the ability to drill down into structs/dicts/composite objects is crucial for a nice experience, but that doesn't have to be at odds with a table based variables viewer. Juno's workspace view for example combines both approaches:
![image](https://user-images.githubusercontent.com/6735977/110916383-066a8f80-8319-11eb-9289-a2de39962d3c.png)

Haven't looked at the proposed tables API yet, but would be cool if table cells could support non-primitives like tree views. Also, does the tables API handle lazy-loading big datasets?

**Comment by davidanthoff** [comment]
So I've been wondering what to do about that :)

I agree with @pfitzseb  that having drill down is _really_ crucial for us, including lazy loading of that etc. I think overall I'm fairly happy with the existing workspace UI we have, even though a table/tree combo a la Juno would of course be even nicer.

I do think that having a shared UI for this would make sense, because clearly this is something that a lot of extensions need/want. But at the same time I'm not sure where it should live :) I think the best option would actually be if VS Code itself had UI for a combined kernel/variable viewer, in the same way that it has UI for notebooks. I think having such a shared UI in the Jupyter extension would probably not be ideal, because for us we would also like to have that UI in scenarios that have essentially nothing to do with a Jupyter scenario... Maybe another option would be to have a "Data Science" VS Code extension? It could add UI for a kernel/variables explorer, a table viewer, a plot viewer, and then expose APIs for extensions like Julia and Jupyter to hook into that and reuse those? We currently _have_ UI for all of these things in the Julia extension, but in many ways I almost feel it would be nicer if there was a shared story for that. And if that is a success, maybe that stuff could at a later point move into the core VS Code product?

Having said that, I think in the short term (v1) the best solution is probably that we just reuse the existing tree variable explorer in the Julia extension for Julia kernels, I think that should be fairly simple.

**Comment by DonJayamanne** [comment]
>  think in the short term (v1) the best solution is probably that we just reuse the existing tree variable ex

Agreed. 

> hink the best option would actually be if VS Code itself had UI for a combined kernel/variable viewer, in the same way that it has UI for notebooks

Agreed.

Also, i didn't see any code in existing Notebook implementation that deals with variables, i guess we'll need to plug that piece of functionality.
I suppose re-using the same repl.ts stuff would make this a possibility (haven't checked the code yet).

I'll close this issue, as its not required right now & i agree with the responses.
